File: android/app/src/main/kotlin/com/elnemr/language/Mp4Chapters.kt
package com.elnemr.language

import android.content.Context
import android.util.Log
import java.io.EOFException
import java.io.RandomAccessFile
import java.security.SecureRandom
import java.security.cert.X509Certificate
import javax.net.ssl.SSLContext
import javax.net.ssl.X509TrustManager
import jcifs.smb.SmbFile
import jcifs.smb.SmbRandomAccessFile
import okhttp3.OkHttpClient
import okhttp3.Request

/// Nero `moov/udta/chpl` chapter reader. Media3 does not surface MP4 chapter
/// atoms, so the player parses them itself: scan top-level boxes for `moov`,
/// then `udta` inside it, then `chpl` inside that. HandBrake and others write
/// `chpl` as Nero chapters (version/flags + count + entries with 8-byte start
/// in 100ns units + 1-byte title length + title). The start is converted to
/// milliseconds (/10000) to match the MKV chapter model.
///
/// Supports local files (RandomAccessFile), SMB (SmbRandomAccessFile), and
/// WebDAV/generic HTTP MP4s (Range: bytes=0-8M → ByteArrayReader). Best-effort:
/// any structural surprise yields whatever was collected so far and never
/// affects playback. Reuses [MkvChapters.Chapter] for emission so the Dart
/// event map stays uniform.
internal object Mp4Chapters {

    class Chapter(
        val title: String,
        val startMs: Long,
        val endMs: Long?,
    )

    private const val MAX_CHAPTERS = 999
    private const val MAX_BOX_HEADER = 16

    fun parse(path: String): List<Chapter> {
        val raf = try {
            RandomAccessFile(path, "r")
        } catch (_: Exception) {
            return emptyList()
        }
        try {
            return doParse(RafReader(raf))
        } catch (e: Exception) {
            Log.i("Mp4Chapters", "parse failed (${e.javaClass.simpleName}: ${e.message})")
            return emptyList()
        } finally {
            try { raf.close() } catch (_: Exception) {}
        }
    }

    fun parseSmb(smbUri: String, context: Context): List<Chapter> {
        val raf = try {
            openSmbFile(smbUri, context) ?: return emptyList()
        } catch (_: Exception) {
            return emptyList()
        }
        try {
            return doParse(SmbReader(raf))
        } catch (e: Exception) {
            Log.i("Mp4Chapters", "parseSmb failed (${e.javaClass.simpleName}: ${e.message})")
            return emptyList()
        } finally {
            try { raf.close() } catch (_: Exception) {}
        }
    }

    fun parseHttp(
        url: String,
        headers: Map<String, String>,
        allowSelfSigned: Boolean,
    ): List<Chapter> {
        return try {
            val bytes = fetchHttpHeader(url, headers, allowSelfSigned) ?: return emptyList()
            doParse(ByteArrayReader(bytes))
        } catch (e: Exception) {
            Log.i("Mp4Chapters", "parseHttp failed (${e.javaClass.simpleName}: ${e.message})")
            emptyList()
        }
    }

    fun parseFtp(ftpUri: String, context: Context): List<Chapter> {
        val src = FtpSeekSource.open(ftpUri, context) ?: return emptyList()
        try {
            return doParse(object : SeekableReader {
                private var p = 0L
                override fun read(): Int {
                    val buf = ByteArray(1)
                    return if (src.readFullyAt(p, buf)) (buf[0].toInt() and 0xFF).also { p += 1 } else -1
                }
                override fun readFully(buf: ByteArray) {
                    if (!src.readFullyAt(p, buf)) throw EOFException()
                    p += buf.size
                }
                override fun seek(pos: Long) { p = pos }
                override fun pos(): Long = p
                override fun length(): Long = src.size
            })
        } catch (e: Exception) {
            Log.i("Mp4Chapters", "parseFtp failed (${e.javaClass.simpleName}: ${e.message})")
            return emptyList()
        } finally {
            src.close()
        }
    }

    private fun fetchHttpHeader(
        url: String,
        headers: Map<String, String>,
        allowSelfSigned: Boolean,
    ): ByteArray? {
        val client = if (allowSelfSigned) permissiveClient else standardClient
        val reqBuilder = Request.Builder().url(url).header("Range", "bytes=0-8388607")
        for ((k, v) in headers) reqBuilder.header(k, v)
        val resp = client.newCall(reqBuilder.build()).execute()
        try {
            if (!resp.isSuccessful && resp.code != 206) return null
            val body = resp.body ?: return null
            val bytes = body.bytes()
            return if (bytes.size > 8 * 1024 * 1024) bytes.copyOf(8 * 1024 * 1024) else bytes
        } finally {
            resp.close()
        }
    }

    private val standardClient: OkHttpClient by lazy { OkHttpClient() }

    private val permissiveClient: OkHttpClient by lazy {
        val trustAll = object : X509TrustManager {
            override fun checkClientTrusted(chain: Array<X509Certificate>, authType: String) = Unit
            override fun checkServerTrusted(chain: Array<X509Certificate>, authType: String) = Unit
            override fun getAcceptedIssuers(): Array<X509Certificate> = arrayOf()
        }
        val sslContext = SSLContext.getInstance("TLS")
        sslContext.init(null, arrayOf(trustAll), SecureRandom())
        OkHttpClient.Builder()
            .sslSocketFactory(sslContext.socketFactory, trustAll)
            .hostnameVerifier { _, _ -> true }
            .build()
    }

    private fun openSmbFile(smbUri: String, context: Context): SmbRandomAccessFile? {
        val uri = android.net.Uri.parse(smbUri) ?: return null
        if (!"smb".equals(uri.scheme, ignoreCase = true)) return null
        val serverId = uri.host ?: return null
        val segments = uri.pathSegments
        if (segments.isEmpty()) return null
        val share = segments[0]
        val remotePath = if (segments.size > 1) segments.subList(1, segments.size).joinToString("/") else ""
        val creds = SmbStore.resolve(context, serverId) ?: return null
        val base = "smb://${creds.host}:${creds.port}/$share"
        val url = if (remotePath.isEmpty()) base else "$base/$remotePath"
        val f = SmbFile(url, creds.context())
        return SmbRandomAccessFile(f, "r")
    }

    // ---- Seekable reader abstraction ----

    private interface SeekableReader {
        fun read(): Int
        fun readFully(buf: ByteArray)
        fun seek(pos: Long)
        fun pos(): Long
        fun length(): Long
    }

    private class RafReader(private val raf: RandomAccessFile) : SeekableReader {
        override fun read(): Int = raf.read()
        override fun readFully(buf: ByteArray) = raf.readFully(buf)
        override fun seek(pos: Long) = raf.seek(pos)
        override fun pos(): Long = raf.filePointer
        override fun length(): Long = try { raf.length() } catch (_: Exception) { Long.MAX_VALUE }
    }

    private class SmbReader(private val raf: SmbRandomAccessFile) : SeekableReader {
        override fun read(): Int = raf.read()
        override fun readFully(buf: ByteArray) = raf.readFully(buf)
        override fun seek(pos: Long) = raf.seek(pos)
        override fun pos(): Long = raf.filePointer
        override fun length(): Long = try { raf.length() } catch (_: Exception) { Long.MAX_VALUE }
    }

    private class ByteArrayReader(private val data: ByteArray) : SeekableReader {
        private var p = 0
        override fun read(): Int = if (p >= data.size) -1 else (data[p++].toInt() and 0xFF)
        override fun readFully(buf: ByteArray) {
            if (p + buf.size > data.size) throw EOFException()
            System.arraycopy(data, p, buf, 0, buf.size)
            p += buf.size
        }
        override fun seek(pos: Long) { p = pos.coerceIn(0, data.size.toLong()).toInt() }
        override fun pos(): Long = p.toLong()
        override fun length(): Long = data.size.toLong()
    }

    private fun doParse(r: SeekableReader): List<Chapter> {
        val fileSize = r.length()
        // Find moov at top level.
        val moov = findBox(r, "moov", 0L, fileSize) ?: return emptyList()
        // Inside moov, find udta. Fallback: search chpl directly under moov.
        val udta = findBox(r, "udta", moov.first, moov.second)
        val chplRange = if (udta != null) {
            findBox(r, "chpl", udta.first, udta.second) ?: findBox(r, "chpl", moov.first, moov.second)
        } else {
            findBox(r, "chpl", moov.first, moov.second)
        } ?: return emptyList()
        val out = parseChpl(r, chplRange.first, chplRange.second)
        finalizeEnds(out)
        return out
    }

    private fun findBox(r: SeekableReader, target: String, start: Long, end: Long): Pair<Long, Long>? {
        val targetBytes = target.toByteArray(Charsets.US_ASCII)
        var pos = start
        while (pos + 8 <= end) {
            r.seek(pos)
            val sizeBytes = ByteArray(4)
            try { r.readFully(sizeBytes) } catch (_: Exception) { break }
            val size = ((sizeBytes[0].toLong() and 0xFF) shl 24) or
                ((sizeBytes[1].toLong() and 0xFF) shl 16) or
                ((sizeBytes[2].toLong() and 0xFF) shl 8) or
                (sizeBytes[3].toLong() and 0xFF)
            val typeBytes = ByteArray(4)
            try { r.readFully(typeBytes) } catch (_: Exception) { break }
            var headerSize = 8L
            var boxSize = size
            if (boxSize == 1L) {
                if (pos + 16 > end) break
                val largeBytes = ByteArray(8)
                try { r.readFully(largeBytes) } catch (_: Exception) { break }
                var v = 0L
                for (b in largeBytes) v = (v shl 8) or (b.toLong() and 0xFF)
                boxSize = v
                headerSize = 16L
            } else if (boxSize == 0L) {
                boxSize = end - pos
            }
            if (boxSize < headerSize) break
            val payloadStart = pos + headerSize
            val payloadEnd = pos + boxSize
            if (payloadEnd > end || payloadEnd < payloadStart) break
            if (typeBytes.contentEquals(targetBytes)) {
                return Pair(payloadStart, payloadEnd)
            }
            pos = payloadEnd
            // Safety: avoid infinite loop on malformed size
            if (pos <= start) break
        }
        return null
    }

    private fun parseChpl(r: SeekableReader, start: Long, end: Long): MutableList<Chapter> {
        val payloadSize = end - start
        if (payloadSize < 5) return mutableListOf()
        r.seek(start)
        val version = r.read()
        if (version < 0) return mutableListOf()
        // flags
        r.read(); r.read(); r.read()
        var headerConsumed = 4L
        if (version != 0) {
            if (payloadSize < headerConsumed + 4 + 1) return mutableListOf()
            r.read(); r.read(); r.read(); r.read()
            headerConsumed += 4
        }
        if (payloadSize < headerConsumed + 1) return mutableListOf()
        val firstCountByte = r.read()
        if (firstCountByte < 0) return mutableListOf()
        headerConsumed += 1

        // Primary: 1-byte count (FFmpeg mov_read_chpl)
        val count1 = firstCountByte
        var best: MutableList<Chapter>? = null
        if (count1 in 1..MAX_CHAPTERS) {
            val parsed = tryParseEntries(r, count1, end)
            if (parsed != null && parsed.isNotEmpty()) best = parsed
        } else if (count1 == 0 && payloadSize > headerConsumed) {
            // count 0 with remaining payload suggests alternative encoding; try fallbacks
        }

        // Fallback 1: count is 32-bit BE where firstCountByte is the high byte
        if ((best == null || best.isEmpty()) && payloadSize >= headerConsumed - 1 + 4) {
            r.seek(start + headerConsumed - 1)
            val b4 = ByteArray(4)
            try {
                r.readFully(b4)
                val count32 = ((b4[0].toLong() and 0xFF) shl 24) or
                    ((b4[1].toLong() and 0xFF) shl 16) or
                    ((b4[2].toLong() and 0xFF) shl 8) or
                    (b4[3].toLong() and 0xFF)
                if (count32 in 1..MAX_CHAPTERS.toLong()) {
                    val parsed = tryParseEntries(r, count32.toInt(), end)
                    if (parsed != null && parsed.isNotEmpty()) best = parsed
                }
            } catch (_: Exception) {}
        }

        // Fallback 2: count is 64-bit BE (8 bytes)
        if ((best == null || best.isEmpty()) && payloadSize >= headerConsumed - 1 + 8) {
            r.seek(start + headerConsumed - 1)
            val b8 = ByteArray(8)
            try {
                r.readFully(b8)
                var v = 0L
                for (b in b8) v = (v shl 8) or (b.toLong() and 0xFF)
                if (v in 1..MAX_CHAPTERS.toLong()) {
                    val parsed = tryParseEntries(r, v.toInt(), end)
                    if (parsed != null && parsed.isNotEmpty()) best = parsed
                }
            } catch (_: Exception) {}
        }

        return best ?: mutableListOf()
    }

    private fun tryParseEntries(r: SeekableReader, count: Int, boxEnd: Long): MutableList<Chapter>? {
        if (count <= 0 || count > MAX_CHAPTERS) return null
        val out = mutableListOf<Chapter>()
        for (i in 0 until count) {
            if (r.pos() + 9 > boxEnd) return null
            val startBytes = ByteArray(8)
            try { r.readFully(startBytes) } catch (_: Exception) { return null }
            var start100ns = 0L
            for (b in startBytes) start100ns = (start100ns shl 8) or (b.toLong() and 0xFF)
            val titleLen = r.read()
            if (titleLen < 0) return null
            if (r.pos() + titleLen > boxEnd) return null
            val titleBytes = ByteArray(titleLen)
            if (titleLen > 0) {
                try { r.readFully(titleBytes) } catch (_: Exception) { return null }
            }
            val raw = String(titleBytes, Charsets.UTF_8).trim()
            val title = if (raw.isEmpty()) "Chapter ${i + 1}" else raw
            val startMs = start100ns / 10000L
            out.add(Chapter(title, startMs, null))
            if (out.size > MAX_CHAPTERS) break
        }
        return out
    }

    private fun finalizeEnds(out: MutableList<Chapter>) {
        out.sortBy { it.startMs }
        for (i in out.indices) {
            val c = out[i]
            if (c.endMs == null && i + 1 < out.size) {
                out[i] = Chapter(c.title, c.startMs, out[i + 1].startMs)
            }
        }
    }
}

File: ios/Runner/Mp4Chapters.swift
import Foundation

/// Nero `moov/udta/chpl` chapter reader (Swift port of `Mp4Chapters.kt`).
/// Media3 / AetherEngine don't surface the Nero `chpl` atom, so we parse the
/// ISO BMFF box tree ourselves: top-level `moov` → `udta` → `chpl`.
///
/// HandBrake writes `chpl` as: version(1) + flags(3) + [4 bytes if version!=0]
/// + count(1 byte, FFmpeg semantics) + entries{ start(8 BE, 100ns) + titleLen(1)
/// + title }. Start is converted to milliseconds (/10000) to match
/// `MkvChapters.Chapter`. A 4-byte and 8-byte count fallback is tried when the
/// 1-byte path yields nothing (handles non-conformant muxers).
enum Mp4Chapters {

    struct Chapter {
        let title: String
        let startMs: Int64
        let endMs: Int64?
    }

    private static let maxChapters = 999

    // MARK: - Public

    static func parse(path: String) -> [Chapter] {
        guard let fh = FileHandle(forReadingAtPath: path) else { return [] }
        defer { try? fh.close() }
        let r = FileReader(handle: fh)
        do {
            return try doParse(r)
        } catch {
            return []
        }
    }

    static func parseMaps(path: String) -> [[String: Any]] {
        parse(path: path).map { c in
            var m: [String: Any] = ["title": c.title, "startMs": c.startMs]
            if let e = c.endMs { m["endMs"] = e }
            return m
        }
    }

    // MARK: - Seekable abstraction

    private protocol SeekableReader {
        func readByte() throws -> Int
        func readFully(int count: Int) throws -> Data
        func seek(to pos: UInt64) throws
        var position: UInt64 { get }
        var length: UInt64 { get }
    }

    private final class FileReader: SeekableReader {
        let handle: FileHandle
        private var pos: UInt64 = 0
        private let fileLength: UInt64
        init(handle: FileHandle) {
            self.handle = handle
            // `seekToEnd` is deprecated but still works; `seekToEndOfFile` is older.
            // Prefer `FileHandle` extensions that don't throw on Linux.
            var len: UInt64 = UInt64.max
            do {
                let cur = try handle.offset()
                let end = try handle.seekToEnd()
                try handle.seek(toOffset: cur)
                len = end
            } catch { len = UInt64.max }
            fileLength = len
        }
        func readByte() throws -> Int {
            let data = try handle.read(upToCount: 1) ?? Data()
            if data.isEmpty { return -1 }
            pos += 1
            return Int(data[0])
        }
        func readFully(int count: Int) throws -> Data {
            let data = try handle.read(upToCount: count) ?? Data()
            if data.count < count { throw NSError(domain: "Mp4Chapters", code: 1) }
            pos += UInt64(count)
            return data
        }
        func seek(to newPos: UInt64) throws {
            try handle.seek(toOffset: newPos)
            pos = newPos
        }
        var position: UInt64 { pos }
        var length: UInt64 { fileLength }
    }

    // MARK: - Parser

    private static func doParse(_ r: SeekableReader) throws -> [Chapter] {
        let fileSize = r.length
        guard let moov = try findBox(r, target: "moov", start: 0, end: fileSize) else { return [] }
        let udta = try findBox(r, target: "udta", start: moov.0, end: moov.1)
        let chpl: (UInt64, UInt64)?
        if let udta {
            chpl = try findBox(r, target: "chpl", start: udta.0, end: udta.1)
                ?? findBox(r, target: "chpl", start: moov.0, end: moov.1)
        } else {
            chpl = try findBox(r, target: "chpl", start: moov.0, end: moov.1)
        }
        guard let chplRange = chpl else { return [] }
        var out = try parseChpl(r, start: chplRange.0, end: chplRange.1)
        finalizeEnds(&out)
        return out
    }

    private static func findBox(_ r: SeekableReader, target: String, start: UInt64, end: UInt64) throws -> (UInt64, UInt64)? {
        let targetBytes = Array(target.utf8)
        guard targetBytes.count == 4 else { return nil }
        var pos = start
        while pos + 8 <= end {
            try r.seek(to: pos)
            let sizeData = try r.readFully(int: 4)
            let typeData = try r.readFully(int: 4)
            var boxSize = (UInt64(sizeData[0]) << 24) | (UInt64(sizeData[1]) << 16) | (UInt64(sizeData[2]) << 8) | UInt64(sizeData[3])
            var headerSize: UInt64 = 8
            if boxSize == 1 {
                if pos + 16 > end { break }
                let large = try r.readFully(int: 8)
                var v: UInt64 = 0
                for b in large { v = (v << 8) | UInt64(b) }
                boxSize = v
                headerSize = 16
            } else if boxSize == 0 {
                boxSize = end - pos
            }
            if boxSize < headerSize { break }
            let payloadStart = pos + headerSize
            let payloadEnd = pos + boxSize
            if payloadEnd > end || payloadEnd < payloadStart { break }
            if typeData[0] == targetBytes[0] && typeData[1] == targetBytes[1] &&
                typeData[2] == targetBytes[2] && typeData[3] == targetBytes[3] {
                return (payloadStart, payloadEnd)
            }
            pos = payloadEnd
            if pos <= start { break }
        }
        return nil
    }

    private static func parseChpl(_ r: SeekableReader, start: UInt64, end: UInt64) throws -> [Chapter] {
        let payloadSize = Int64(end) - Int64(start)
        if payloadSize < 5 { return [] }
        try r.seek(to: start)
        let version = try r.readByte()
        if version < 0 { return [] }
        // flags 3 bytes
        _ = try r.readByte(); _ = try r.readByte(); _ = try r.readByte()
        var headerConsumed: Int64 = 4
        if version != 0 {
            if payloadSize < headerConsumed + 4 + 1 { return [] }
            _ = try r.readByte(); _ = try r.readByte(); _ = try r.readByte(); _ = try r.readByte()
            headerConsumed += 4
        }
        if payloadSize < headerConsumed + 1 { return [] }
        let firstCountByte = try r.readByte()
        if firstCountByte < 0 { return [] }
        headerConsumed += 1
        let count1 = firstCountByte
        var best: [Chapter]? = nil
        if (1...maxChapters).contains(count1) {
            if let parsed = try tryParseEntries(r, count: count1, boxEnd: end) , !parsed.isEmpty {
                best = parsed
            }
        }
        // Fallback: 32-bit count where firstCountByte is high byte
        if (best == nil || best!.isEmpty), payloadSize >= headerConsumed - 1 + 4 {
            try r.seek(to: start + UInt64(headerConsumed - 1))
            let b4 = try r.readFully(int: 4)
            let count32 = (UInt64(b4[0]) << 24) | (UInt64(b4[1]) << 16) | (UInt64(b4[2]) << 8) | UInt64(b4[3])
            if (1...UInt64(maxChapters)).contains(count32) {
                if let parsed = try tryParseEntries(r, count: Int(count32), boxEnd: end), !parsed.isEmpty {
                    best = parsed
                }
            }
        }
        // Fallback: 64-bit count
        if (best == nil || best!.isEmpty), payloadSize >= headerConsumed - 1 + 8 {
            try r.seek(to: start + UInt64(headerConsumed - 1))
            let b8 = try r.readFully(int: 8)
            var v: UInt64 = 0
            for b in b8 { v = (v << 8) | UInt64(b) }
            if (1...UInt64(maxChapters)).contains(v) {
                if let parsed = try tryParseEntries(r, count: Int(v), boxEnd: end), !parsed.isEmpty {
                    best = parsed
                }
            }
        }
        return best ?? []
    }

    private static func tryParseEntries(_ r: SeekableReader, count: Int, boxEnd: UInt64) throws -> [Chapter]? {
        if count <= 0 || count > maxChapters { return nil }
        var out: [Chapter] = []
        for i in 0..<count {
            if r.position + 9 > boxEnd { return nil }
            let startData = try r.readFully(int: 8)
            var start100ns: UInt64 = 0
            for b in startData { start100ns = (start100ns << 8) | UInt64(b) }
            let titleLen = try r.readByte()
            if titleLen < 0 { return nil }
            if r.position + UInt64(titleLen) > boxEnd { return nil }
            var title = "Chapter \(i + 1)"
            if titleLen > 0 {
                let titleData = try r.readFully(int: titleLen)
                let raw = String(data: titleData, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
                if !raw.isEmpty { title = raw }
            }
            let startMs = Int64(start100ns / 10000)
            out.append(Chapter(title: title, startMs: startMs, endMs: nil))
            if out.count > maxChapters { break }
        }
        return out
    }

    private static func finalizeEnds(_ out: inout [Chapter]) {
        out.sort { $0.startMs < $1.startMs }
        for i in out.indices {
            if out[i].endMs == nil, i + 1 < out.count {
                let c = out[i]
                out[i] = Chapter(title: c.title, startMs: c.startMs, endMs: out[i+1].startMs)
            }
        }
    }
}
